### PR TITLE
use nonblocking tun fd again

### PR DESCRIPTION
**Pull request by wongsyrone**
pumping data in golang native world

Signed-off-by: Syrone Wong <wong.syrone@gmail.com>

**NOTE:** Need https://github.com/trojan-gfw/igniter-go-libs/pull/3

### CODE REVIEW

File: app/src/main/java/io/github/trojan_gfw/igniter/ProxyService.java [hunk 1]
```diff
@@ -185,38 +151,13 @@ private void shutdown() {
             Log.e("Clash", "clash stopped");
         }
         Tun2socks.stop();
```
Comment (oasiscifr) [inline app/src/main/java/io/github/trojan_gfw/igniter/ProxyService.java]: Is `Tun2socks.stop();` a block or non-block call? Is it possible to cause the UI reenabled while the `stopping` is still  in progress?

Comment (oasiscifr) [reply app/src/main/java/io/github/trojan_gfw/igniter/ProxyService.java]: Moreover, the tun device could be closed by system(tun off VPN in system setting). Is this situation considered?

File: app/src/main/java/io/github/trojan_gfw/igniter/ProxyService.java [hunk 2]
```diff
@@ -185,38 +151,13 @@ private void shutdown() {
...
-            Log.e(TAG, Thread.currentThread().getName() + " thread start");
-            while (state == STARTING || state == STARTED) {
-                try {
-                    int n = inputStream.read(packetBuffer.array());
-                    if (n > 0) {
-                        packetBuffer.limit(n);
-                        Tun2socks.inputPacket(packetBuffer.array());
-                        packetBuffer.clear();
-                    }
-                } catch (IOException e) {
```
Comment (oasiscifr) [inline app/src/main/java/io/github/trojan_gfw/igniter/ProxyService.java]: The tun device could be closed by system(tun off VPN in system setting). The catch here can handle it. Is this situation considered in your design?

File: app/src/main/java/io/github/trojan_gfw/igniter/ProxyService.java (hunk 1 above)
Comment (wongsyrone) [reply app/src/main/java/io/github/trojan_gfw/igniter/ProxyService.java]: A blocking call. The button is disabled while the stopping is in progress.

File: app/src/main/java/io/github/trojan_gfw/igniter/ProxyService.java (hunk 2 above)
Comment (wongsyrone) [reply app/src/main/java/io/github/trojan_gfw/igniter/ProxyService.java]: https://developer.android.com/reference/android/net/VpnService#onRevoke()

The issue should be solved via commit "close resources on revoke"

File: app/src/main/java/io/github/trojan_gfw/igniter/ProxyService.java (hunk 1 above)
Comment (oasiscifr) [reply app/src/main/java/io/github/trojan_gfw/igniter/ProxyService.java]: Oh, the correct words are "sync call" and "async call", lol

**Comment by oasiscifr** [comment]
[app-release.apk.zip](https://github.com/trojan-gfw/igniter/files/4009631/app-release.apk.zip)

**Pull request merged**